src/main/java/com/google/sps/servlets/RecommendServlet.java

```diff
@@ -0,0 +1,159 @@
+      recommendations = datastore.get(recKey);
+      List<Long> recIds = (List<Long>) recommendations.getProperty("recs");
+      int limit = EVENTS_LIMIT;
+      for (int i = 0; i < recIds.size() && i < limit; i++) {
+        long eventId = recIds.get(i);
+        Key eventKey = KeyFactory.createKey("Event", eventId);
+        try {
+          events.add(datastore.get(eventKey));
+        } catch (EntityNotFoundException eventNotFound) {
+          LOGGER.info("event " + eventId + " not found.");
+          limit++;
+        }
+      }
+    } catch (EntityNotFoundException | NullPointerException exception) {
+      // no prev recommendations, compute simple recommendations based on tags and distance only
+      LOGGER.info("No recommendations found for " + userID + ". Computing now.");
+      recommendations = new Entity("Recommendation", userID);
+      Key userKey = KeyFactory.createKey("User", userID);
+      Entity userEntity = null;
+      try {
+        userEntity = datastore.get(userKey);
+        Map<String, Float> userParams = Interactions.buildVectorForUser(userEntity);
+        if (userParams.size() > 0) {
+          // compute recommendation if user has data
+          String userLocation = null;
+          if (userEntity.hasProperty("location")) {
+            userLocation = userEntity.getProperty("location").toString();
+          }
+          Map<Double, Long> bestEvents = new TreeMap<>(Recommend.SCORE_DESCENDING);
+          Map<Long, Entity> eventsWithId = new HashMap<>();
+          for (Entity event : datastore.prepare(new Query("Event")).asIterable()) {
+            long eventId = event.getKey().getId();
+            eventsWithId.put(eventId, event);
+            Map<String, Integer> eventParams = Interactions.buildVectorForEvent(event);
+            String eventLocation = event.getProperty("address").toString();
+            double eventScore =
+                computeScore(userID, userParams, userLocation, eventId, eventParams, eventLocation);
+            Recommend.addToRanking(eventId, eventScore, bestEvents);
+          }
+          List<Long> eventIds = new ArrayList<>();
+          Iterator<Double> itr = bestEvents.keySet().iterator();
+          int count = 0;
+          while (itr.hasNext() && count < EVENTS_LIMIT) {
+            count++;
+            long eventId = bestEvents.get(itr.next());
+            events.add(eventsWithId.get(eventId));
+            eventIds.add(eventId);
+          }
+          recommendations = new Entity("Recommendation", userID);
+          recommendations.setProperty("recs", eventIds);
+          datastore.put(recommendations);
+        }
+      } catch (EntityNotFoundException userNotFound) {
+        // user does not exist (no data)
+        userEntity = Utils.makeUserEntity(userID, true);
+      }
+    }
+    response.getWriter().println(gson.toJson(events));
+  }
+
+  /** Computes the correlation between a user and an event. */
+  private static double computeScore(
```

Comment:
This method is hard to understand. Please add docstring to explain its behaviour.

Reply:
Done!